lib/new_relic/agent/instrumentation/ruby_openai/instrumentation.rb

```diff
@@ -0,0 +1,175 @@
+      ensure
+        finish(segment, event)
+      end
+    end
+
+    def chat_completions_instrumentation(parameters)
+      segment = NewRelic::Agent::Tracer.start_segment(name: CHAT_COMPLETIONS_SEGMENT_NAME)
+      record_openai_metric
+      event = create_chat_completion_summary(parameters)
+      segment.llm_event = event
+      messages = create_chat_completion_messages(parameters, event.id)
+
+      begin
+        response = NewRelic::Agent::Tracer.capture_segment_error(segment) { yield }
+        # TODO: Remove !response.include?('error) when we drop support for versions below 4.0.0
+        if response && !response.include?('error')
+          add_chat_completion_response_params(parameters, response, event)
+          messages = update_chat_completion_messages(messages, response, event)
+        end
+
+        response
+      ensure
+        finish(segment, event)
+        messages&.each { |m| m.record }
+      end
+    end
+
+    def create_chat_completion_summary(parameters)
+      NewRelic::Agent::Llm::ChatCompletionSummary.new(
+        # TODO: POST-GA: Add metadata from add_custom_attributes if prefixed with 'llm.', except conversation_id
+        vendor: VENDOR,
+        conversation_id: conversation_id,
+        api_key_last_four_digits: parse_api_key,
+        request_max_tokens: parameters[:max_tokens] || parameters['max_tokens'],
+        request_model: parameters[:model] || parameters['model'],
+        temperature: parameters[:temperature] || parameters['temperature']
+      )
+    end
+
+    def create_embeddings_event(parameters)
+      NewRelic::Agent::Llm::Embedding.new(
+        # TODO: POST-GA: Add metadata from add_custom_attributes if prefixed with 'llm.', except conversation_id
+        vendor: VENDOR,
+        input: parameters[:input] || parameters['input'],
+        api_key_last_four_digits: parse_api_key,
+        request_model: parameters[:model] || parameters['model']
+      )
+    end
+
+    def add_chat_completion_response_params(parameters, response, event)
+      event.response_number_of_messages = (parameters[:messages] || parameters['messages']).size + response['choices'].size
+      # The response hash always returns keys as strings, so we don't need to run an || check here
+      event.response_model = response['model']
+      event.response_usage_total_tokens = response['usage']['total_tokens']
+      event.response_usage_prompt_tokens = response['usage']['prompt_tokens']
+      event.response_usage_completion_tokens = response['usage']['completion_tokens']
+      event.response_choices_finish_reason = response['choices'][0]['finish_reason']
+    end
+
+    def add_embeddings_response_params(response, event)
+      event.response_model = response['model']
+      event.response_usage_total_tokens = response['usage']['total_tokens']
+      event.response_usage_prompt_tokens = response['usage']['prompt_tokens']
+    end
+
+    def parse_api_key
+      'sk-' + headers['Authorization'][-4..-1]
```

Comment:
Hmm... it looks like the ability to use `[-4..]` instead of `[-4..-1]` wasn't introduced until Ruby v2.6.